### PR TITLE
Add operations support

### PR DESCRIPTION
**Pull request by acaron**
Make possible to add `operations.rb` in a concept without a model. This implementation also make `form_for` and `simple_form_for` work with `@form` even if there's no model associated. Support for controller active_record is added as a `monkey patch` since it is deprecated.

### CODE REVIEW

**Comment by apotonick** [comment]
Thaaaaanks! :heart: 

1. So is there a new convention coming up? `crud.rb` if there's a model to load prior to the operation, `operations.rb` if not? I didn't even think about that. Is that how you use it?
2. The way you changed responder irritates me. Is there any use cases where you would use a resopnder without a model? Aren't they all written against the "ActiveModel API"? 

Please don't mix 3 things into one PR in the future.

**Comment by acaron** [comment]
What are the 3 different things? Everything is needed to make it works!

Right now yes it is crud with model and operations without! The best case would be always operations and then you decide if it needs a model or not... Since you can operation with model outside the conventional crud.

We can remove Responder without model completely! I'm might misunderstood how to use it with the method errors, etc.

**Comment by acaron** [comment]
What about:
1. Remove Responder outside model;
2. Deprecate crud.rb;
3. Keep operations.rb but add model loading only when Model is included;
4. Add some tests!

**Comment by apotonick** [comment]
Haha the `persisted?` is rendering-only. If you don't use form renderers, for example, you don't need it.

How would we detect whether or not a model is "included" in an operation? Some folks don't use the `Model` module.

**Comment by acaron** [comment]
If they don't use the Model module, they will not have any of the delegates, etc. So @model is never set so autoloading is not needed? They still have access to the models directory in Rails. Do I miss something? It's hard to follow an already huge code! (All the trb suites and gems)

Yep the form rendering of course but I tried to cover everything needed :grin:

**Comment by timoschilling** [comment]
I think this PR has changes for more then one PR, too.

You should add a better description of your intend for this PR.

The `crud` pattern should be removed and the `operations` pattern should be default, for Trb in general.
For me I now use the pattern `Model::Operations::Action` and `Model::Cells::Action`.

**Comment by apotonick** [comment]
Ah, I think I understand our miscommunication now! Loading the model file has nothing to do with delegations in the first place. The main problem is that we have to load the model file when we put operations into that model's namespace. If we don't load the model file, the whole app breaks.

The delegations, making it work with simpleform, etc. are things that have to be solved differently. `Model` has nothing to do with ActiveModel/ActiveRecord, it's more like "this operation uses some persistent object and now that you configured it using `model` it knows how to find/create that particular "model". Don't confuse that with any Simpleform-`persisted?`-whatever stuff. Does it make it a bit clearer?

**Comment by apotonick** [comment]
Here's a quick example what we're trying to solve.

```ruby
# app/concepts/comment/operations.rb

class Comment < ActiveRecord::Base # this defines the model class, and does NOT load the app/models/comment.rb file!
  class Create < Comment::Operation
```

The easiest solution for all that is as follows.


```ruby
# app/concepts/comment/operations.rb

require_dependency "app/models/comment"
class Comment < ActiveRecord::Base
  class Create < Comment::Operation
```

I think an explicit list where we configure our loading will be the simplest solution. The other things (responder, `persisted?`, etc.) are different problems, @acaron that we should solve in a separate PR. :beers:

**Comment by acaron** [comment]
Yes, I will uncommit the persisted thing soon and move it to another branch lol

**Comment by acaron** [comment]
Just remove the form feature to keep only the `operations.rb` integration. I suggest:
1. Make `crud.rb` deprecated;
2. Add to the documentation that `require_dependency` must be added until we find a solution as discussed. I do mine in an initializer so it is done automatically.

**Comment by timoschilling** [comment]
There is a other pattern with don't need the `require_dependency`.

```
app
├── concepts
│   ├── comment
│   │   ├── cell.rb
│   │   ├── views
│   │   │   ├── show.haml
│   │   │   ├── list.haml
│   │   ├── assets
│   │   │   ├── comment.css.sass
│   │   ├── operations
│   │   │   ├── create.rb
│   │   │   ├── update.rb
│   │   ├── twin.rb
```

**Comment by acaron** [comment]
Can you explain? I don't understand what you are trying to say. The `require_dependency` is for the model you use from the Rails `models/` directory only.

**Comment by timoschilling** [comment]
If you name your operations like `Post::Operations::Create` you can work with the normal Rails autoloading.

`require_dependency` is only a Rails wrapper for `require` which re-require (re-load) the file on change.

This is the pseudo code of `require_dependency`:
```ruby
$dependencies = []
def require_dependency
end
```

**Comment by acaron** [comment]
If my model use by my operations are not in `require_dependency`, it makes my application goes weird (because I have a `Scope` lib in a model that is not reload if a file change for example which is an issue). Makes everything pretty instable.

This PR doesn't do: `require_dependency "#{app.root}/app/models/#{model}"` in the method `autoload_operations` since some operations in `operations.rb` can be modeless. But it must be added manually if you use `operations.rb` or `create.rb`, etc. with Rails models until we found another solution to autoload them.

We can't autoload them right now because model are chosen with `model` method inside an operation for some user. We can autoload them if the user is using `self.model_class` override instead.

**Comment by timoschilling** [comment]
Trust me there is no need for the `require_dependency` usage.

**Comment by acaron** [comment]
You must explain how because, if I use external scopes in my Rails model, I can't call the scope from the operation. For example, in my model I have: `include Scopes::Translations` it throws `undefined method 'translations_with_fallback'` if I don't require my models.

Even if I don't use `require_dependency`, I need at least a `require` (like you said, `require_dependency` is a container) which lead to the same problem that it must be done manually.

**Comment by timoschilling** [comment]
Try this:
```ruby
# models/post.rb
class Post < AR::Base; end
```
```ruby
# concepts/post/operations/create.rb
class Post::Operations::Create < Trp::Operation; end  
```
What's happen if you call `Post::Operations::Create.run` in your controller is:
1. the constant `Post::Operations::Create` is missing, so it will be converted to `post/operations/create.rb` and would be required
2. `Post` is missing and would be loaded form `models/post.rb`
3. `Post::Operation` is missing and can't be loaded, so it would be create by the autoloader
4. `Post::Operation::Create` is defined by the file
Every thing is done without the `require_dependency`, `require_dependency` is only needed if you don't use the Rails way. Keep in mind "don't fight the framework" :wink:

**Comment by acaron** [comment]
Ahhh that is a solution I like! We can remove all the require part in Railtie and only add `module Operations` instead. I will test it to make sure it works and I will change the Railtie file.

**Comment by acaron** [comment]
Thanks @timoschilling, the solution was simple (stupid). I was searching way too far for autoloading thing. I will wait for @apotonick approval.

So:
1. Deprecate `crud.rb`;
2. Put your operations inside `Operations` module instead. Call the full path to work with Rails autoloading.

**Comment by apotonick** [comment]
Can we split this PR into a few smaller ones? I'm losing track of what's supposed to go on here.

@acaron The `Operation::Model` module is _not_ meant to have any coupling to ActiveRecord. The `persisted?` etc. is highly ActiveRecord-specific. Please don't put that into the core code. :stuck_out_tongue_winking_eye: As I said, imagine all this is soon to be used in another framework like Lotus or Roda, where we don't use SimpleForm.

The same goes with `Responder`. Again, this is super Rails-specific, which is why I put it into a separate module. This won't be around forever, so save your time on this module!!! :smile: 

Both problems you're experiencing come from the presentation layer in Rails (SimpleForm and responders are both presentation). We shall not pollute our business layer with presentation code, just to make that stupid form renderer work. It might look a bit redundant, but the plan is to slowly segregate from Rails and all its guessing.

Maybe the confusion also comes from `Model`. As I said, this is not to be understood like "ActiveRecord _model_", this could be any model, e.g. a ROM or Lotus::Model instance.

**Comment by acaron** [comment]
persisted has been remove! Look at the final files change

**Comment by acaron** [comment]
Actually, the operation is very couple with a model... That is why I move Rails specific model things when the module Model is included only... Sane as persisted, it looks like a form and an operation must have a model

**Comment by apotonick** [comment]
An operation is coupled to a model when it decides to do so. Including `Model` is not enough, unfortunately. `Model` uses a very generic interface, it basically only requires the configured "model" class to provide `Comment::find` and `Comment::new`, which will work with any ORM right now, or can easily be adaptered (not sure if that's a word haha).

The methods you added are highly specific to Rails and it's way of guessing about model internals. Let's put it the other way: The `Model` module for operation is in no way meant to be limited to Rails models, it should work with any major ORM. Your change will blur that a bit, as it now looks as if it's made for ActiveRecord, only. Does that make it a bit better to understand?

The main goal of this PR is to allow `operations.rb`, right?

**Comment by apotonick** [comment]
Oh, and something else I've noticed!!! By removing the `Responder` module you do not allow modelless operations being passed into `respond_to`! :grimacing: The responder interface is completely different from the way forms are rendered, even though they seem to use the same methods.

Imagine you had a modelless operation that you want to be presented by a responder - you now have to mix in `Model` to make that work. Hahaha, complicated, I know! :stuck_out_tongue_winking_eye:

**Comment by acaron** [comment]
Ok, so put back Responder and test `repond_to` and yes it is the main goal! What would you changes in the code exactly?
